File: dotnet/JBMClient/util/SimpleString.cs
/*
 * JBoss, Home of Professional Open Source
 * Copyright 2005-2008, Red Hat Middleware LLC, and individual contributors
 * by the @authors tag. See the copyright.txt in the distribution for a
 * full listing of individual contributors.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */


using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace JBoss.JBM.Client.util
{
    /**
     * 
     * A SimpleString
     * 
     * A simple String class that can store all characters, and stores as simple byte[],
     * this minimises expensive copying between String objects
     * 
     * @author <a href="mailto:tim.fox@jboss.com">Tim Fox</a>
     * @author <a href="mailto:clebert.suconic@jboss.com">Clebert Suconic</a>
     *
     */
    public class SimpleString : IComparable
    {
        // Attributes
        // ------------------------------------------------------------------------
        private byte[] data;

        private int hash;

        // Cache the string
        private string str;

        // Static
        // ----------------------------------------------------------------------

        /**
         * Returns a SimpleString constructed from the <code>string</code> parameter.
         * If <code>string</code> is <code>null</code>, the return value will be <code>null</code> too.
         */
        public static SimpleString ToSimpleString(string value)
        {
            if (value == null)
            {
                return null;
            }
            return new SimpleString(value);
        }

        // Constructors
        // ----------------------------------------------------------------------

        public SimpleString(string value)
        {

            int j = 0;

            char[] characters = value.ToCharArray();

            int len = characters.Length;

            data = new byte[len << 1];

            for (int i = 0; i < len; i++)
            {
                char c = characters[i];

                byte low = (byte)(c & 0xFF); // low byte

                data[j++] = low;

                byte high = (byte)(c >> 8 & 0xFF); // high byte

                data[j++] = high;
            }

            str = value;
        }

        public SimpleString(byte[] data)
        {
            this.data = data;
        }

        // CharSequence implementation
        // ---------------------------------------------------------------------------

        public int Length
        {
            get
            {
                return data.Length >> 1;
            }
        }

        public char CharAt(int pos)
        {
            if (pos < 0 || pos >= data.Length >> 1)
            {
                throw new IndexOutOfRangeException();
            }
            pos <<= 1;

            return (char)(data[pos] | data[pos + 1] << 8);
        }

        public SimpleString SubSequence(int start, int end)
        {
            int len = data.Length >> 1;

            if (end < start || start < 0 || end > len)
            {
                throw new IndexOutOfRangeException();
            }
            else
            {
                int newlen = (end - start) << 1;
                byte[] bytes = new byte[newlen];

                Array.Copy(data, start << 1, bytes, 0, newlen);

                return new SimpleString(bytes);
            }
        }

        // Comparable implementation -------------------------------------

        public int CompareTo(Object o)
        {
            return this.ToString().CompareTo(o.ToString());
        }

        // Public
        // ---------------------------------------------------------------------------

		public byte[] Data
		{
			get
			{
				return this.data;
			}
			private set
			{
				this.data = value;
			}
		}

        public bool StartsWith(SimpleString other)
        {
            byte[] otherdata = other.data;

            if (otherdata.Length > this.data.Length)
            {
                return false;
            }

            for (int i = 0; i < otherdata.Length; i++)
            {
                if (this.data[i] != otherdata[i])
                {
                    return false;
                }
            }

            return true;
        }


        public override string ToString()
        {
            if (str == null)
            {
                int len = data.Length >> 1;

                char[] chars = new char[len];

                int j = 0;

                for (int i = 0; i < len; i++)
                {
                    int low = data[j++] & 0xFF;

                    int high = (data[j++] << 8) & 0xFF00;

                    chars[i] = (char)(low | high);
                }

                str = new String(chars);
            }

            return str;
        }

        public override bool Equals(Object other)
        {
            if (other is SimpleString)
            {
                SimpleString s = (SimpleString)other;

                if (data.Length != s.data.Length)
                {
                    return false;
                }

                for (int i = 0; i < data.Length; i++)
                {
                    if (data[i] != s.data[i])
                    {
                        return false;
                    }
                }

                return true;
            }
            else
            {
                return false;
            }
        }

        public override int GetHashCode()
        {
            if (hash == 0)
            {
                for (int i = 0; i < data.Length; i++)
                {
                    hash = (31 * hash + data[i]) % (int.MaxValue>>5);
                }
            }

            return hash;
        }

        public SimpleString[] Split(char delim)
        {
            if (!Contains(delim))
            {
                return new SimpleString[] { this };
            }
            else
            {
                IList<SimpleString> all = new List<SimpleString>();
                int lasPos = 0;
                for (int i = 0; i < data.Length; i += 2)
                {
                    byte low = (byte)(delim & 0xFF); // low byte
                    byte high = (byte)(delim >> 8 & 0xFF); // high byte
                    if (data[i] == low && data[i + 1] == high)
                    {
                        byte[] bytes = new byte[i - lasPos];
                        Array.Copy(data, lasPos, bytes, 0, bytes.Length);
                        lasPos = i + 2;
                        all.Add(new SimpleString(bytes));
                    }
                }
                byte[] bytes2 = new byte[data.Length - lasPos];
                Array.Copy(data, lasPos, bytes2, 0, bytes2.Length);
                all.Add(new SimpleString(bytes2));
                return all.ToArray();
            }
        }

        public bool Contains(char c)
        {
            for (int i = 0; i < data.Length; i += 2)
            {
                byte low = (byte)(c & 0xFF); // low byte
                byte high = (byte)(c >> 8 & 0xFF); // high byte
                if (data[i] == low && data[i + 1] == high)
                {
                    return true;
                }
            }
            return false;
        }

        public SimpleString Concat(SimpleString toAdd)
        {

            byte[] bytes = new byte[data.Length + toAdd.Data.Length];
            Array.Copy(data, 0, bytes, 0, data.Length);
            Array.Copy(toAdd.Data, 0, bytes, data.Length, toAdd.Data.Length);
            return new SimpleString(bytes);
        }

        public SimpleString Concat(char c)
        {
            byte[] bytes = new byte[data.Length + 2];
            Array.Copy(data, 0, bytes, 0, data.Length);
            bytes[data.Length] = (byte)(c & 0xFF);
            bytes[data.Length + 1] = (byte)(c >> 8 & 0xFF);
            return new SimpleString(bytes);
        }

        public static int SizeofString(SimpleString str)
        {

            return DataConstants.SIZE_INT + str.data.Length;
        }

        public static int SizeofNullableString(SimpleString str)
        {
            if (str == null)
            {
                return 1;
            }
            else
            {
                return 1 + SizeofString(str);
            }
        }

        public void GetChars(int srcBegin, int srcEnd, char[] dst, int dstBegin)
        {
            if (srcBegin < 0)
            {
                throw new IndexOutOfRangeException("Invalid initialIndex: " + srcBegin);
            }
            if (srcEnd > this.Length)
            {
                throw new IndexOutOfRangeException("Invalid endIndex: " + srcEnd);
            }
            if (srcBegin > srcEnd)
            {
                throw new IndexOutOfRangeException("Invalid Length: " + (srcEnd - srcBegin));
            }

            int j = 0;

            for (int i = srcBegin; i < srcEnd - srcBegin; i++)
            {
                int low = data[j++] & 0xFF;

                int high = (data[j++] << 8) & 0xFF00;

                dst[i] = (char)(low | high);
            }
        }

    }
}

File: src/main/org/hornetq/utils/SimpleString.java
/*
 * Copyright 2009 Red Hat, Inc.
 * Red Hat licenses this file to you under the Apache License, version
 * 2.0 (the "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *    http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied.  See the License for the specific language governing
 * permissions and limitations under the License.
 */

package org.hornetq.utils;

import static org.hornetq.utils.DataConstants.SIZE_INT;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

import org.hornetq.core.logging.Logger;

/**
 * 
 * A SimpleString
 * 
 * A simple String class that can store all characters, and stores as simple byte[],
 * this minimises expensive copying between String objects
 * 
 * @author <a href="mailto:tim.fox@jboss.com">Tim Fox</a>
 * 
 * TODO - implement an intern() method like in java.lang.String, since many Strings e.g. addresses, queue names, remote node ids are duplicated heavily
 * in bindings taking up more memory than they should
 * Intern can be called when receiving a sent message at the server (destination)
 * Also when receiving bindings remotely via bridge, the address, queue name and node id can be interned
 *
 */
public class SimpleString implements CharSequence, Serializable, Comparable<SimpleString>
{
   private static final long serialVersionUID = 4204223851422244307L;

   private static final Logger log = Logger.getLogger(SimpleString.class);

   // Attributes
   // ------------------------------------------------------------------------
   private final byte[] data;

   private transient int hash;

   // Cache the string
   private transient String str;

   // Static
   // ----------------------------------------------------------------------

   /**
    * Returns a SimpleString constructed from the <code>string</code> parameter.
    * If <code>string</code> is <code>null</code>, the return value will be <code>null</code> too.
    */
   public static SimpleString toSimpleString(final String string)
   {
      if (string == null)
      {
         return null;
      }
      return new SimpleString(string);
   }

   // Constructors
   // ----------------------------------------------------------------------

   public SimpleString(final String string)
   {
      int len = string.length();

      data = new byte[len << 1];

      int j = 0;

      for (int i = 0; i < len; i++)
      {
         char c = string.charAt(i);

         byte low = (byte)(c & 0xFF); // low byte

         data[j++] = low;

         byte high = (byte)(c >> 8 & 0xFF); // high byte

         data[j++] = high;
      }

      str = string;
   }
   
   public SimpleString(final byte[] data)
   {
      this.data = data;
   }

   // CharSequence implementation
   // ---------------------------------------------------------------------------

   public int length()
   {
      return data.length >> 1;
   }

   public char charAt(int pos)
   {
      if (pos < 0 || pos >= data.length >> 1)
      {
         throw new IndexOutOfBoundsException();
      }
      pos <<= 1;

      return (char)(data[pos] | data[pos + 1] << 8);
   }

   public CharSequence subSequence(final int start, final int end)
   {
      int len = data.length >> 1;

      if (end < start || start < 0 || end > len)
      {
         throw new IndexOutOfBoundsException();
      }
      else
      {
         int newlen = (end - start) << 1;
         byte[] bytes = new byte[newlen];

         System.arraycopy(data, start << 1, bytes, 0, newlen);

         return new SimpleString(bytes);
      }
   }

   // Comparable implementation -------------------------------------

   public int compareTo(SimpleString o)
   {
      return toString().compareTo(o.toString());
   }

   // Public
   // ---------------------------------------------------------------------------

   public byte[] getData()
   {
      return data;
   }

   public boolean startsWith(final SimpleString other)
   {
      byte[] otherdata = other.data;

      if (otherdata.length > this.data.length)
      {
         return false;
      }

      for (int i = 0; i < otherdata.length; i++)
      {
         if (this.data[i] != otherdata[i])
         {
            return false;
         }
      }

      return true;
   }

   public String toString()
   {
      if (str == null)
      {
         int len = data.length >> 1;

         char[] chars = new char[len];

         int j = 0;

         for (int i = 0; i < len; i++)
         {
            int low = data[j++] & 0xFF;

            int high = (data[j++] << 8) & 0xFF00;

            chars[i] = (char)(low | high);
         }

         str = new String(chars);
      }

      return str;
   }

   public boolean equals(Object other)
   {
      if (other instanceof SimpleString)
      {
         SimpleString s = (SimpleString)other;

         if (data.length != s.data.length)
         {
            return false;
         }

         for (int i = 0; i < data.length; i++)
         {
            if (data[i] != s.data[i])
            {
               return false;
            }
         }

         return true;
      }
      else
      {
         return false;
      }
   }

   public int hashCode()
   {
      if (hash == 0)
      {
         int tmphash = 0;
         for (int i = 0; i < data.length; i++)
         {
            tmphash = (tmphash << 5) - tmphash + data[i]; // (hash << 5) - hash is same as hash * 31
         }
         hash = tmphash;
      }

      return hash;
   }
   
   public SimpleString[] split(char delim)
   {
      if (!contains(delim))
      {
         return new SimpleString[] { this };
      }
      else
      {
         List<SimpleString> all = new ArrayList<SimpleString>();
         int lasPos = 0;
         for (int i = 0; i < data.length; i += 2)
         {
            byte low = (byte)(delim & 0xFF); // low byte
            byte high = (byte)(delim >> 8 & 0xFF); // high byte
            if (data[i] == low && data[i + 1] == high)
            {
               byte[] bytes = new byte[i - lasPos];
               System.arraycopy(data, lasPos, bytes, 0, bytes.length);
               lasPos = i + 2;
               all.add(new SimpleString(bytes));
            }
         }
         byte[] bytes = new byte[data.length - lasPos];
         System.arraycopy(data, lasPos, bytes, 0, bytes.length);
         all.add(new SimpleString(bytes));
         SimpleString[] parts = new SimpleString[all.size()];
         return all.toArray(parts);
      }
   }

   public boolean contains(char c)
   {
      for (int i = 0; i < data.length; i += 2)
      {
         byte low = (byte)(c & 0xFF); // low byte
         byte high = (byte)(c >> 8 & 0xFF); // high byte
         if (data[i] == low && data[i + 1] == high)
         {
            return true;
         }
      }
      return false;
   }
   
   public SimpleString concat(final String toAdd)
   {
      return concat(new SimpleString(toAdd));
   }

   public SimpleString concat(final SimpleString toAdd)
   {
      byte[] bytes = new byte[data.length + toAdd.getData().length];
      System.arraycopy(data, 0, bytes, 0, data.length);
      System.arraycopy(toAdd.getData(), 0, bytes, data.length, toAdd.getData().length);
      return new SimpleString(bytes);
   }

   public SimpleString concat(final char c)
   {
      byte[] bytes = new byte[data.length + 2];
      System.arraycopy(data, 0, bytes, 0, data.length);
      bytes[data.length] = (byte)(c & 0xFF);
      bytes[data.length + 1] = (byte)(c >> 8 & 0xFF);
      return new SimpleString(bytes);
   }

   public int sizeof()
   {
      return SIZE_INT + data.length;
   }
   
   public static int sizeofString(final SimpleString str)
   {
      return str.sizeof();
   }

   public static int sizeofNullableString(final SimpleString str)
   {
      if (str == null)
      {
         return 1;
      }
      else
      {
         return 1 + str.sizeof();
      }
   }

   public void getChars(int srcBegin, int srcEnd, char dst[], int dstBegin)
   {
      if (srcBegin < 0)
      {
         throw new StringIndexOutOfBoundsException(srcBegin);
      }
      if (srcEnd > length())
      {
         throw new StringIndexOutOfBoundsException(srcEnd);
      }
      if (srcBegin > srcEnd)
      {
         throw new StringIndexOutOfBoundsException(srcEnd - srcBegin);
      }

      int j = 0;

      for (int i = srcBegin; i < srcEnd - srcBegin; i++)
      {
         int low = data[j++] & 0xFF;

         int high = (data[j++] << 8) & 0xFF00;

         dst[i] = (char)(low | high);
      }
   }

}